chain/builder.go

```diff
@@ -387,62 +395,73 @@ func BuildBlock(
-		return nil, fmt.Errorf("%w: unable to insert height", err)
+	if err := tsv.Insert(ctx, heightKey, binary.BigEndian.AppendUint64(nil, b.height)); err != nil {
+		return nil, nil, nil, fmt.Errorf("%w: unable to insert height", err)
 	}
-	if err := tsv.Insert(ctx, timestampKey, binary.BigEndian.AppendUint64(nil, uint64(b.Tmstmp))); err != nil {
-		return nil, fmt.Errorf("%w: unable to insert timestamp", err)
+	if err := tsv.Insert(ctx, timestampKey, binary.BigEndian.AppendUint64(nil, uint64(b.timestamp))); err != nil {
+		return nil, nil, nil, fmt.Errorf("%w: unable to insert timestamp", err)
 	}
 	if err := tsv.Insert(ctx, feeKey, feeManager.Bytes()); err != nil {
-		return nil, fmt.Errorf("%w: unable to insert fees", err)
+		return nil, nil, nil, fmt.Errorf("%w: unable to insert fees", err)
 	}
 	tsv.Commit()
 
 	// Fetch [parentView] root as late as possible to allow
 	// for async processing to complete
 	root, err := parentView.GetMerkleRoot(ctx)
 	if err != nil {
-		return nil, err
+		return nil, nil, nil, err
 	}
-	b.StateRoot = root
+	b.stateRoot = root
 
 	// Get view from [tstate] after writing all changed keys
-	view, err := ts.ExportMerkleDBView(ctx, vm.Tracer(), parentView)
+	view, err := ts.ExportMerkleDBView(ctx, c.tracer, parentView)
 	if err != nil {
-		return nil, err
+		return nil, nil, nil, err
 	}
 
-	// Compute block hash and marshaled representation
-	if err := b.initializeBuilt(ctx, view, results, feeManager); err != nil {
-		log.Warn("block failed", zap.Int("txs", len(b.Txs)), zap.Any("consumed", feeManager.UnitsConsumed()))
-		return nil, err
+	// Initialize finalized metadata fields
+	blk, err := NewStatelessBlock(
+		b.parentID,
+		b.timestamp,
+		b.height,
+		b.txs,
+		b.stateRoot,
+	)
+	if err != nil {
+		return nil, nil, nil, err
 	}
 
 	// Kickoff root generation
 	go func() {
 		start := time.Now()
 		root, err := view.GetMerkleRoot(ctx)
 		if err != nil {
-			log.Error("merkle root generation failed", zap.Error(err))
+			c.log.Error("merkle root generation failed", zap.Error(err))
 			return
 		}
-		log.Info("merkle root generated",
-			zap.Uint64("height", b.Hght),
-			zap.Stringer("blkID", b.ID()),
+		c.log.Info("merkle root generated",
+			zap.Uint64("height", blk.Hght),
+			zap.Stringer("blkID", blk.ID()),
 			zap.Stringer("root", root),
 		)
-		b.vm.RecordRootCalculated(time.Since(start))
+		c.metrics.rootCalculated.Observe(float64(time.Since(start)))
 	}()
 
-	log.Info(
+	c.log.Info(
 		"built block",
-		zap.Uint64("hght", b.Hght),
+		zap.Uint64("hght", b.height),
 		zap.Int("attempted", txsAttempted),
-		zap.Int("added", len(b.Txs)),
+		zap.Int("added", len(b.txs)),
 		zap.Int("state changes", ts.PendingChanges()),
 		zap.Int("state operations", ts.OpIndex()),
 		zap.Int64("parent (t)", parent.Tmstmp),
-		zap.Int64("block (t)", b.Tmstmp),
+		zap.Int64("block (t)", b.timestamp),
 	)
-	return b, nil
+	return NewExecutionBlockNoVerify(blk), &ExecutedBlock{
```

Comment:
It feels weird to me that we return both the execution+executed block, can't we just return the executed block result?

Reply:
The block is the actual container being executed and the result is for APIs